File: sdk/go/commands.go
package literedis

import (
	"fmt"
	"time"
)

func (c *Client) Set(key string, value interface{}, expiration time.Duration) error {
	var args []interface{}
	args = append(args, key, value)
	if expiration > 0 {
		args = append(args, "PX", int64(expiration/time.Millisecond))
	}
	_, err := c.Do("SET", args...)
	if err != nil {
		return fmt.Errorf("SET command failed: %w", err)
	}
	return nil
}

func (c *Client) Get(key string) (string, error) {
	reply, err := c.Do("GET", key)
	if err != nil {
		return "", fmt.Errorf("GET command failed: %w", err)
	}
	if reply == nil {
		return "", nil
	}
	return reply.(string), nil
}

func (c *Client) Del(keys ...string) (int64, error) {
	args := make([]interface{}, len(keys))
	for i, key := range keys {
		args[i] = key
	}
	reply, err := c.Do("DEL", args...)
	if err != nil {
		return 0, fmt.Errorf("DEL command failed: %w", err)
	}
	return reply.(int64), nil
}

func (c *Client) Incr(key string) (int64, error) {
	reply, err := c.Do("INCR", key)
	if err != nil {
		return 0, fmt.Errorf("INCR command failed: %w", err)
	}
	return reply.(int64), nil
}

func (c *Client) Expire(key string, expiration time.Duration) (bool, error) {
	reply, err := c.Do("EXPIRE", key, int64(expiration/time.Second))
	if err != nil {
		return false, fmt.Errorf("EXPIRE command failed: %w", err)
	}
	return reply.(int64) == 1, nil
}

func (c *Client) TTL(key string) (time.Duration, error) {
	reply, err := c.Do("TTL", key)
	if err != nil {
		return 0, fmt.Errorf("TTL command failed: %w", err)
	}
	ttl := reply.(int64)
	if ttl < 0 {
		return -1, nil
	}
	return time.Duration(ttl) * time.Second, nil
}

func (c *Client) MSet(pairs map[string]interface{}) error {
	args := make([]interface{}, 0, len(pairs)*2)
	for k, v := range pairs {
		args = append(args, k, v)
	}
	_, err := c.Do("MSET", args...)
	if err != nil {
		return fmt.Errorf("MSET command failed: %w", err)
	}
	return nil
}

func (c *Client) MGet(keys ...string) ([]interface{}, error) {
	args := make([]interface{}, len(keys))
	for i, key := range keys {
		args[i] = key
	}
	reply, err := c.Do("MGET", args...)
	if err != nil {
		return nil, fmt.Errorf("MGET command failed: %w", err)
	}
	return reply.([]interface{}), nil
}

File: sdk/rust/src/commands.rs
use crate::client::AsyncClient;
use crate::error::RedisError;
use crate::resp::Value;

#[async_trait::async_trait]
pub trait Commands: AsyncClient {
    async fn set(&mut self, key: &str, value: &str) -> Result<(), RedisError> {
        let command = format!("*3\r\n$3\r\nSET\r\n${}\r\n{}\r\n${}\r\n{}\r\n", 
                              key.len(), key, value.len(), value);
        match self.execute(command.as_bytes()).await? {
            Value::SimpleString(s) if s == "OK" => Ok(()),
            _ => Err(RedisError::ProtocolError("Unexpected response".to_string())),
        }
    }

    async fn get(&mut self, key: &str) -> Result<Option<String>, RedisError> {
        let command = format!("*2\r\n$3\r\nGET\r\n${}\r\n{}\r\n", key.len(), key);
        match self.execute(command.as_bytes()).await? {
            Value::BulkString(Some(bytes)) => Ok(Some(String::from_utf8(bytes)?)),
            Value::BulkString(None) => Ok(None),
            _ => Err(RedisError::ProtocolError("Unexpected response".to_string())),
        }
    }

  async fn del(&mut self, keys: &[&str]) -> Result<i64, RedisError> {
        let mut command = format!("*{}\r\n$3\r\nDEL\r\n", keys.len() + 1);
        for key in keys {
            command.push_str(&format!("${}\r\n{}\r\n", key.len(), key));
        }
        match self.execute(command.as_bytes()).await? {
            Value::Integer(n) => Ok(n),
            _ => Err(RedisError::ProtocolError("Unexpected response".to_string())),
        }
    }

    async fn incr(&mut self, key: &str) -> Result<i64, RedisError> {
        let command = format!("*2\r\n$4\r\nINCR\r\n${}\r\n{}\r\n", key.len(), key);
        match self.execute(command.as_bytes()).await? {
            Value::Integer(n) => Ok(n),
            _ => Err(RedisError::ProtocolError("Unexpected response".to_string())),
        }
    }

    async fn expire(&mut self, key: &str, seconds: u64) -> Result<bool, RedisError> {
        let command = format!("*3\r\n$6\r\nEXPIRE\r\n${}\r\n{}\r\n${}\r\n{}\r\n",
                              key.len(), key, seconds.to_string().len(), seconds);
        match self.execute(command.as_bytes()).await? {
            Value::Integer(1) => Ok(true),
            Value::Integer(0) => Ok(false),
            _ => Err(RedisError::ProtocolError("Unexpected response".to_string())),
        }
    }

    async fn ttl(&mut self, key: &str) -> Result<i64, RedisError> {
        let command = format!("*2\r\n$3\r\nTTL\r\n${}\r\n{}\r\n", key.len(), key);
        match self.execute(command.as_bytes()).await? {
            Value::Integer(n) => Ok(n),
            _ => Err(RedisError::ProtocolError("Unexpected response".to_string())),
        }
    }

    async fn set_ex(&mut self, key: &str, value: &str, expiration: Duration) -> Result<(), RedisError> {
        let seconds = expiration.as_secs();
        let command = format!("*4\r\n$5\r\nSETEX\r\n${}\r\n{}\r\n${}\r\n{}\r\n${}\r\n{}\r\n",
                              key.len(), key, seconds.to_string().len(), seconds, value.len(), value);
        match self.execute(command.as_bytes()).await? {
            Value::SimpleString(s) if s == "OK" => Ok(()),
            _ => Err(RedisError::ProtocolError("Unexpected response".to_string())),
        }
    }
}

impl Commands for crate::client::Client {}
